ENGAGEHF/Dashboard/Vitals/VitalsManager.swift

```diff
@@ -0,0 +1,353 @@
+//
+// This source file is part of the ENGAGE-HF project based on the Stanford Spezi Template Application project
+//
+// SPDX-FileCopyrightText: 2024 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import FirebaseAuth
+import FirebaseFirestore
+import Foundation
+import HealthKit
+import struct ModelsR4.DateTime
+import class ModelsR4.ObservationComponent
+import OSLog
+import Spezi
+import SpeziFirebaseConfiguration
+
+
+/// Vitals History Manager
+///
+/// Functionality:
+/// - Maintain local, up-to-date arrays of the patients health data
+/// - Convert FHIR quantities to HK Samples
+/// - Manage units of the stored quantities, defaulting to the user's phone settings
+@Observable
+public class VitalsManager: Module, EnvironmentAccessible {
+    enum VitalsError: Error {
+        case invalidConversion
+        case unknownUnit
+        case invalidObservationType
+        case missingField
+    }
+    
+    
+    @ObservationIgnored @Dependency private var configureFirebaseApp: ConfigureFirebaseApp
+    @ObservationIgnored @StandardActor private var standard: ENGAGEHFStandard
+    private let logger = Logger(subsystem: "ENGAGEHF", category: "VitalsManager")
+    
+    private var authStateDidChangeListenerHandle: AuthStateDidChangeListenerHandle?
+    private var snapshotListener: ListenerRegistration?
+    
+    // TODO: Localize the units based on either locale or user preferences in the phone settings
+    public var heartRateHistory: [HKQuantitySample] = []
+    public var bloodPressureHistory: [HKCorrelation] = []
+    public var weightHistory: [HKQuantitySample] = []
+    
+    
+    public var latestHeartRate: HKQuantitySample? {
+        heartRateHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestBloodPressure: HKCorrelation? {
+        bloodPressureHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestWeight: HKQuantitySample? {
+        weightHistory.max { $0.startDate < $1.startDate }
+    }
+    
+    
+    /// Call on initial configuration:
+    /// - Add a snapshot listener to the three health data collections
+    /// - Set the default units to the user's localized preferences
+    public func configure() {
+        if ProcessInfo.processInfo.isPreviewSimulator {
+            self.setupPreview()
+            return
+        }
+        
+        authStateDidChangeListenerHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
+            self?.registerSnapshotListeners(user: user)
+        }
+        
+        self.registerSnapshotListeners(user: Auth.auth().currentUser)
+    }
+    
+    private func registerSnapshotListeners(user: User?) {
+        self.logger.debug("Initializing vitals snapshot listener...")
+        
+        // Remove previous snapshot listener for the user before creating new one
+        snapshotListener?.remove()
```

Comment:
We do never store anything in the `snapshotListener`. It might make sense to make this an array, store each of the snapshot listeners in there and then map on all of them here, call remove, and then clear out the array.